apps/omg/lib/omg/state/transaction/validator.ex

```diff
@@ -14,74 +14,49 @@
   @moduledoc """
-  Provides functions for stateful transaction validation for transaction processing in OMG.State.Core.
+  Dispatches validation flow to payments application or fee claiming modules
   """
 
-  @maximum_block_size 65_536
+  require OMG.State.Transaction.Payment
+
+  # NOTE: Last processed transaction could potentially take his room but also generate `max_inputs` fee transactions
+  @safety_margin 2 * OMG.State.Transaction.Payment.max_inputs()
+  @maximum_block_size 65_536 - @safety_margin
 
   alias OMG.Fees
-  alias OMG.Output
   alias OMG.State.Core
   alias OMG.State.Transaction
-  alias OMG.State.UtxoSet
-  alias OMG.Utxo
-
-  require Utxo
-
-  @type exec_error ::
-          :amounts_do_not_add_up
-          | :fees_not_covered
-          | :input_utxo_ahead_of_state
-          | :too_many_transactions_in_block
-          | :unauthorized_spend
-          | :utxo_not_found
+  alias OMG.State.Transaction.Validator
 
-  @spec can_apply_spend(state :: Core.t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
-          true | {{:error, exec_error()}, Core.t()}
-  def can_apply_spend(
-        %Core{utxos: utxos} = state,
-        %Transaction.Recovered{signed_tx: %{raw_tx: raw_tx}, witnesses: witnesses} = tx,
-        fees
-      ) do
-    inputs = Transaction.get_inputs(tx)
+  @type can_process_tx_error ::
+          :too_many_transactions_in_block | Validator.Payment.can_apply_error() | Validator.FeeClaim.fee_claim_error()
 
-    with :ok <- validate_block_size(state),
-         :ok <- inputs_not_from_future_block?(state, inputs),
-         {:ok, outputs_spent} <- UtxoSet.get_by_inputs(utxos, inputs),
-         :ok <- authorized?(outputs_spent, witnesses),
-         {:ok, implicit_paid_fee_by_currency} <- Transaction.Protocol.can_apply?(raw_tx, outputs_spent),
-         true <- Fees.covered?(implicit_paid_fee_by_currency, fees) || {:error, :fees_not_covered} do
-      true
-    else
-      {:error, _reason} = error -> {error, state}
-    end
+  @doc """
+  Checks, whether at a given state of the ledger, a particular transaction can be applied (!) to it,
+  subject to particular fee requirements
+  """
+  @spec can_process_tx(state :: Core.t(), tx :: Transaction.Recovered.t(), fees :: Fees.optional_fee_t()) ::
+          {:ok, :apply_spend, map()} | {:ok, :claim_fees, map()} | {{:error, can_process_tx_error()}, Core.t()}
+  def can_process_tx(%Core{} = state, %Transaction.Recovered{} = tx, fees) do
+    with :ok <- validate_block_size(state), do: dispatch_validation(state, tx, fees)
   end
 
-  defp validate_block_size(%Core{tx_index: number_of_transactions_in_block}) do
-    case number_of_transactions_in_block == @maximum_block_size do
+  defp validate_block_size(%Core{tx_index: number_of_transactions_in_block, fees_paid: fees_paid}) do
+    fee_transactions_count = Enum.count(fees_paid)
+
+    case number_of_transactions_in_block + fee_transactions_count >= @maximum_block_size do
```

Comment:
this might sound like a nitpick, but if `Enum.count(fees_paid) == 1` and there is a single spot "left" in the block, and the last transaction to hit this pays in a distinct token (making `count(fees_paid) == 2`) it looks like it will bork.

So you should prepare 2 slots and possibly "waste" one if `count` remains at 1

Would be nice to have a unit test for this :sweat_smile: